neural_modelling/src/neuron/plasticity/stdp/synapse_dynamics_stdp_mad_neuromodulated_impl.c

```diff
@@ -0,0 +1,455 @@
+// Spinn_common includes
+#include "static-assert.h"
+
+// sPyNNaker neural modelling includes
+#include "../../synapses.h"
+
+// Plasticity common includes
+#include "../common/maths.h"
+#include "../common/post_events.h"
+
+#include "weight_dependence/weight.h"
+#include "timing_dependence/timing.h"
+#include <string.h>
+#include <debug.h>
+
+uint32_t num_plastic_pre_synaptic_events = 0;
+
+//---------------------------------------
+// Macros
+//---------------------------------------
+// The plastic control words used by Morrison synapses store an axonal delay in the upper 3 bits
+// Assuming a maximum of 16 delay slots, this is all that is required as:
+//
+// 1) Dendritic + Axonal <= 15
+// 2) Dendritic >= Axonal
+//
+// Therefore:
+//
+// * Maximum value of dendritic delay is 15 (with axonal delay of 0) - It requires 4 bits
+// * Maximum value of axonal delay is 7 (with dendritic delay of 8) - It requires 3 bits
+//
+//             |        Axonal delay       |  Dendritic delay   |       Type        |      Index         |
+//             |---------------------------|--------------------|-------------------|--------------------|
+// Bit count   | SYNAPSE_AXONAL_DELAY_BITS | SYNAPSE_DELAY_BITS | SYNAPSE_TYPE_BITS | SYNAPSE_INDEX_BITS |
+//             |                           |                    |        SYNAPSE_TYPE_INDEX_BITS         |
+//             |---------------------------|--------------------|----------------------------------------|
+#ifndef SYNAPSE_AXONAL_DELAY_BITS
+  #define SYNAPSE_AXONAL_DELAY_BITS 3
+#endif
+
+#define SYNAPSE_AXONAL_DELAY_MASK ((1 << SYNAPSE_AXONAL_DELAY_BITS) - 1)
+
+#define SYNAPSE_DELAY_TYPE_INDEX_BITS \
+    (SYNAPSE_DELAY_BITS + SYNAPSE_TYPE_INDEX_BITS)
+
+#if (SYNAPSE_DELAY_TYPE_INDEX_BITS + SYNAPSE_AXONAL_DELAY_BITS) > 16
+  #error "Not enough bits for axonal synaptic delay bits"
+#endif
+
+//---------------------------------------
+// Structures
+//---------------------------------------
+typedef struct
+{
+  pre_trace_t prev_trace;
+  uint32_t prev_time;
+} pre_event_history_t;
+
+post_event_history_t *post_event_history;
+uint32_t weight_update_constant_component;
+uint32_t last_dopamine_spike_time = 0;
+weight_state_t weight_state;
+
+
+//---------------------------------------
+// Dopamine trace is a simple decaying trace similarly implemented as pre and
+// post trace.
+static inline post_trace_t add_dopamine_spike(
+        uint32_t time, int16_t concentration, int16_t* last_dopamine_level) {
```

Comment:
Again, passing 16-bit integers to functions is inefficient - C will believe you know what you're doing and mask the results of all the 32-bit ARM operations down to 16-bit